File: code/494/Java/dp2.java
class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        int total = Arrays.stream(nums).sum();
        if (total<S || ((total^S)&1)==1) {
            return 0;
        }
        int plus = (total+S)>>1;
        int cnt[] = new int[plus+1];
        cnt[0] = 1;
        for (int n:nums) {
            for (int i=plus; i>=n; i--) {
                cnt[i] += cnt[i-n];
            }
        }
        return cnt[plus];
    }
}

File: code/494/Python/dp2.py
class Solution:
    def findTargetSumWays(self, nums, S):
        total = sum(nums)
        if total<S or (total^S)&1: return 0
        target = (S+total)>>1
        cnt = [0]*(target+1)
        cnt[0] = 1
        for n in nums:
            for i in range(target, -1, -1):
                if i-n<0: break
                cnt[i] += cnt[i-n]
        return cnt[target]